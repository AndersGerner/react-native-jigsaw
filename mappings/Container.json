{"name":"Container","tag":"Container","description":"A container component with gutter padding","category":"layout","supports_list_render":false,"layout":{},"props":{"useThemeGutterPadding":{"label":"Use gutter padding","description":"When true, uses the theme gutter spacing as the container's horizontal padding","type":"boolean","value":true,"editable":true,"required":true},"backgroundImage":{"label":"Background image","description":"Background image to apply to the container","type":"remoteImage","value":null,"editable":true,"required":false},"backgroundImageResizeMode":{"label":"Background image resize mode","description":"Determines how to resize the background image when the frame doesn't match the raw image dimensions","editable":true,"required":false,"value":null,"type":"flatArray","options":["cover","contain","stretch","repeat","center"]},"elevation":{"label":"Elevation","description":"Elevation of the component. A number 0-3.","type":"flatArray","options":[0,1,2,3],"value":0,"editable":true}}}