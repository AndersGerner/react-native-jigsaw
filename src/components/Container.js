import * as React from "react";
import { View, Image } from "react-native";
import { withTheme } from "../core/theming";
import {
  COMPONENT_TYPES,
  FORM_TYPES,
  BORDER_RADIUS_MODE,
  ELEVATION_TYPE
} from "../core/component-types";
import Elevation from "./Elevation";

class Container extends React.Component {
  render() {
    const {
      theme: { spacing },
      useThemeGutterPadding,
      backgroundColor,
      backgroundImage,
      backgroundImageResizeMode,
      elevation,
      style,
      children
    } = this.props;

    const containerStyle = {
      paddingHorizontal: useThemeGutterPadding ? spacing.gutters : 0,
      backgroundColor,
      width: "100%"
    };

    const Wrap = elevation ? Elevation : View;

    if (elevation) {
      containerStyle.elevation = elevation;
    }

    return (
      <Wrap style={[containerStyle, style]}>
        <React.Fragment>
          {backgroundImage ? (
            <Image
              source={
                typeof backgroundImage === "string"
                  ? { uri: backgroundImage }
                  : backgroundImage
              }
              resizeMode={backgroundImageResizeMode}
              style={{
                position: "absolute",
                top: 0,
                bottom: 0,
                left: 0,
                right: 0
              }}
            />
          ) : null}
          {children}
        </React.Fragment>
      </Wrap>
    );
  }
}

export default withTheme(Container);

export const SEED_DATA = {
  name: "Container",
  tag: "Container",
  description: "A container component with gutter padding",
  type: COMPONENT_TYPES.primitive,
  supports_list_render: false,
  layout: {},
  props: {
    useThemeGutterPadding: {
      label: "Use gutter padding",
      description:
        "When true, uses the theme gutter spacing as the container's horizontal padding",
      type: FORM_TYPES.boolean,
      value: true,
      editable: true,
      required: true
    },
    backgroundColor: {
      label: "Background color",
      description: "Background color to apply to the container",
      type: FORM_TYPES.color,
      value: "background",
      editable: true,
      required: false
    },
    backgroundImage: {
      label: "Background image",
      description: "Background image to apply to the container",
      type: FORM_TYPES.remoteImage,
      value: null,
      editable: true,
      required: false
    },
    backgroundImageResizeMode: {
      label: "Background image resize mode",
      description:
        "Determines how to resize the background image when the frame doesn't match the raw image dimensions",
      editable: true,
      required: false,
      value: null,
      type: FORM_TYPES.flatArray,
      options: ["cover", "contain", "stretch", "repeat", "center"]
    },
    borderRadiusMode: BORDER_RADIUS_MODE,
    elevation: {
      ...ELEVATION_TYPE,
      value: 0
    }
  }
};
