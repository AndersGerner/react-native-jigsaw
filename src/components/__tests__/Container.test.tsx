import * as React from "react";
import renderer from "react-test-renderer";
import Container from "../Container";

it("renders container", () => {
  const tree = renderer
    .create(
      <Container
        backgroundColor="blue"
        borderColor="red"
        borderWidth={2}
        useThemeGutterPadding
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
