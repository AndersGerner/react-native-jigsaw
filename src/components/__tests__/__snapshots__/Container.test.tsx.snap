// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders container 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "blue",
        "borderColor": "red",
        "borderWidth": 2,
        "paddingHorizontal": 16,
        "width": "100%",
      },
      undefined,
    ]
  }
/>
`;
